File: android/src/main/kotlin/de/julianostarek/motif/ui/player/PlayerBar.kt
/*
 * Copyright 2022 Julian Ostarek
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package de.julianostarek.motif.ui.player

import androidx.compose.foundation.Image
import androidx.compose.foundation.clickable
import androidx.compose.foundation.layout.*
import androidx.compose.material.ExperimentalMaterialApi
import androidx.compose.material.ModalBottomSheetState
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.rounded.Pause
import androidx.compose.material.icons.rounded.PlayArrow
import androidx.compose.material.icons.rounded.Warning
import androidx.compose.material3.*
import androidx.compose.runtime.*
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.res.painterResource
import androidx.compose.ui.text.style.TextOverflow
import androidx.compose.ui.unit.dp
import com.google.accompanist.drawablepainter.rememberDrawablePainter
import de.julianostarek.motif.player.RemoteState
import kotlinx.coroutines.launch

@OptIn(ExperimentalMaterialApi::class)
@Composable
fun PlayerBar(
    viewModel: AndroidPlayerViewModel,
    bottomSheetState: ModalBottomSheetState,
    modifier: Modifier = Modifier
) {
    var isPresentingControls by remember { mutableStateOf(false) }
    val state = viewModel.shared.remoteState.collectAsState().value

    Surface(
        shape = MaterialTheme.shapes.medium,
        shadowElevation = 4.dp,
        tonalElevation = 4.dp,
        modifier = modifier.clickable(
            enabled = state is RemoteState.Connected,
            onClick = { isPresentingControls = !isPresentingControls }
        ),
    ) {
        Column {
            Row(
                modifier = Modifier.height(56.dp).padding(8.dp),
                verticalAlignment = Alignment.CenterVertically,
                horizontalArrangement = Arrangement.spacedBy(8.dp)
            ) {
                if (state is RemoteState.Connected) {
                    val imageModifier = Modifier
                        .clip(MaterialTheme.shapes.medium)
                        .fillMaxHeight()
                    val trackImage = viewModel.trackImage.collectAsState().value
                    if (trackImage != null) {
                        Image(
                            painter = rememberDrawablePainter(trackImage),
                            contentDescription = null,
                            modifier = imageModifier,
                        )
                    } else {
                        DefaultCoverArt(modifier = imageModifier)
                    }
                }

                MetadataText(state, modifier = Modifier.weight(1F))
                Spacer(Modifier)

                if (state is RemoteState.Connected) {
                    PlayPauseButton(viewModel, state)
                } else {
                    ConnectButton(bottomSheetState, state)
                }
            }
            if (isPresentingControls) {
                Column {
                    Divider(modifier = Modifier.padding(horizontal = 8.dp))
                    AdditionalControls(viewModel, bottomSheetState, state)
                }
            }
        }
    }

}

@Composable
private fun MetadataText(
    state: RemoteState,
    modifier: Modifier = Modifier
) {
    if (state is RemoteState.Connected) {
        Column(verticalArrangement = Arrangement.SpaceBetween, modifier = modifier) {
            Text(
                text = when (state) {
                    is RemoteState.Connected.Playback -> {
                        "${state.track.title} • ${state.track.artists.joinToString()}"
                    }

                    else -> "No Playback"
                },
                style = MaterialTheme.typography.bodyMedium,
                maxLines = 1,
                overflow = TextOverflow.Ellipsis
            )

            Row(
                verticalAlignment = Alignment.CenterVertically,
                horizontalArrangement = Arrangement.spacedBy(8.dp)
            ) {
                Image(
                    painter = painterResource(state.service.brandingIconRes),
                    contentDescription = null
                )

                CompositionLocalProvider(LocalContentColor provides MaterialTheme.colorScheme.onSurfaceVariant) {
                    Text(
                        text = state.service.brandingName,
                        style = MaterialTheme.typography.labelSmall
                    )
                }
            }
        }
    } else {
        CompositionLocalProvider(LocalContentColor provides MaterialTheme.colorScheme.onSurfaceVariant) {
            Row(
                verticalAlignment = Alignment.CenterVertically,
                horizontalArrangement = Arrangement.spacedBy(8.dp),
                modifier = modifier
            ) {
                Icon(imageVector = Icons.Rounded.Warning, contentDescription = null)
                Text("Not Connected", style = MaterialTheme.typography.bodyMedium)
            }
        }
    }
}

@Composable
fun Enabled(enabled: Boolean, content: @Composable () -> Unit) {
    if (!enabled) {
        CompositionLocalProvider(LocalContentColor provides MaterialTheme.colorScheme.onSurfaceVariant.copy(alpha = 0.33F)) {
            content()
        }
    } else {
        content()
    }
}

@Composable
private fun PlayPauseButton(
    viewModel: AndroidPlayerViewModel,
    state: RemoteState,
    modifier: Modifier = Modifier
) {
    if (state is RemoteState.Connected) {
        val playbackState = state as? RemoteState.Connected.Playback

        Enabled(playbackState != null) {
            IconButton(
                enabled = playbackState != null,
                onClick = {
                    if (playbackState?.isPaused != false) {
                        viewModel.shared.resume()
                    } else {
                        viewModel.shared.pause()
                    }
                },
                modifier = modifier
            ) {
                Icon(
                    imageVector = if (playbackState?.isPaused != false) Icons.Rounded.PlayArrow else Icons.Rounded.Pause,
                    contentDescription = null,
                    modifier = Modifier.size(32.dp)
                )
            }
        }
    }
}

@Composable
private fun OpenButton(
    viewModel: AndroidPlayerViewModel
) {
    TextButton(onClick = {}) {
        Text("Open")
    }
}

@OptIn(ExperimentalMaterialApi::class)
@Composable
private fun ConnectButton(
    bottomSheetState: ModalBottomSheetState,
    state: RemoteState,
    modifier: Modifier = Modifier
) {
    val scope = rememberCoroutineScope()
    TextButton(
        onClick = { scope.launch { bottomSheetState.show() } },
        modifier = modifier
    ) {
        Text(if (state is RemoteState.Connected) "Switch" else "Connect")
    }
}

@OptIn(ExperimentalMaterialApi::class)
@Composable
private fun AdditionalControls(
    viewModel: AndroidPlayerViewModel,
    bottomSheetState: ModalBottomSheetState,
    state: RemoteState,
    modifier: Modifier = Modifier
) {
    Row(modifier = modifier) {
        OpenButton(viewModel)
        ConnectButton(bottomSheetState, state)
    }
}

File: ios/Motif/Player/PlayerBar.swift
//
//  PlayerBar.swift
//
//  Copyright 2022 Julian Ostarek
//
//  Licensed under the Apache License, Version 2.0 (the "License");
//  you may not use this file except in compliance with the License.
//  You may obtain a copy of the License at
//
//      http://www.apache.org/licenses/LICENSE-2.0
//
//  Unless required by applicable law or agreed to in writing, software
//  distributed under the License is distributed on an "AS IS" BASIS,
//  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
//  See the License for the specific language governing permissions and
//  limitations under the License.
//

import Foundation
import SwiftUI
import Shared
import Introspect

struct PlayerBar: View {
    @ObservedObject var viewModel: PlayerViewModelShim
    @State var isPresentingControls: Bool = false

    var body: some View {
        Button {
            withAnimation(.easeInOut(duration: 0.2)) {
                isPresentingControls.toggle()
            }
        } label: {
            VStack {
                HStack(spacing: 12) {
                    if isConnected {
                        Group {
                            if let trackImage = viewModel.trackImage {
                                Image(uiImage: trackImage)
                                    .resizable()
                                    .clipShape(RoundedRectangle(cornerSize: .init(width: 6, height: 6)))
                                    .aspectRatio(contentMode: .fill)
                                    .transition(.opacity.animation(.easeInOut))
                                    .scaledToFit()
                                    .id(viewModel.trackImage)
                            } else {
                                DefaultCoverArt()
                            }
                        }
                        .padding([.leading, .top, .bottom], 8)
                    }

                    metadataAndService
                    Spacer()
                    if isConnected {
                        playPauseButton
                    } else {
                        openButton
                    }
                }
                .frame(height: 56)
                if isPresentingControls {
                    HStack {
                        openButton
                        Button("Switch") {
                            viewModel.isPresentingChooser = true
                        }
                    }
                    .frame(height: 56)
                }
            }
            .padding(.trailing)
            .onOpenURL(perform: {
                viewModel.authorizeFromUrl(url: $0)
            })
            .background(
                RoundedRectangle(cornerSize: .init(width: 6, height: 6))
                    .fill(Color(uiColor: .secondarySystemBackground))
            )
            .shadow(radius: 8)
        }
        .buttonStyle(.plain)
    }

    var isConnected: Bool {
        viewModel.remoteState is Shared.RemoteState.Connected
    }

    var playPauseButton: some View {
        Group {
            if isConnected {
                let playbackState = viewModel.remoteState as? Shared.RemoteState.ConnectedPlayback
                Button {
                    if let isPaused = playbackState?.isPaused,
                       isPaused {
                        viewModel.shared.resume()
                    }else {
                        viewModel.shared.pause()
                    }
                } label: {
                    Image(systemName: (playbackState?.isPaused ?? false) ? "play.fill" : "pause.fill")
                        .resizable()
                        .aspectRatio(contentMode: .fit)
                        .frame(height: 20)
                }
                .buttonStyle(.plain)
                .disabled(playbackState == nil)
            }
        }
    }

    var metadataAndService: some View {
        VStack(alignment: .leading, spacing: 4) {
            if let connectedState = (viewModel.remoteState as? Shared.RemoteState.Connected) {
                let text: String = {
                    if let playbackState = connectedState as? Shared.RemoteState.ConnectedPlayback {
                        return "\(playbackState.track.title) • \(playbackState.track.artists.joined(separator: ", "))"
                    } else {
                        return "No Playback"
                    }
                }()
                Text(text)
                    .font(.callout)
                    .fontWeight(.bold)

                Label {
                    var label: String? = nil
                    switch connectedState.service {
                    case PlayerPlayerService.appleMusic:
                        let _ = label = "Apple Music"
                    case PlayerPlayerService.spotify:
                        let _ = label = "Spotify"
                    default:
                        let _ = ()
                    }
                    if let label {
                        Text(label)
                            .font(.caption)
                            .foregroundColor(.secondary)
                    }
                } icon: {
                    var name: String? = nil
                    switch connectedState.service {
                    case PlayerPlayerService.appleMusic:
                        let _ = name = "AppleMusic"
                    case PlayerPlayerService.spotify:
                        let _ = name = "Spotify"
                    default:
                        let _ = ()
                    }
                    if let name {
                        Image(name)
                            .resizable()
                            .interpolation(.high)
                            .frame(width: 16, height: 16)
                    }
                }
            } else {
                Label("Not Connected", systemImage: "exclamationmark.circle.fill")
                    .foregroundColor(.secondary)
                    .padding(.leading)
            }
        }
    }

    var openButton: some View {
        Group {
            if let connectedState = viewModel.remoteState as? Shared.RemoteState.Connected {
                Button("Open") {

                }
            } else {
                Button("Connect") {
                    viewModel.isPresentingChooser = true
                }
            }
        }
    }
}
